# Проект по SQL

**Заказчик**
Коммерческий департамент компании

**Цель исследования**
 - Анализ базы данных свежекупленного сервиса для формирования концепции нового продукта.

**Задания:**
 - Посчитайте, сколько книг вышло после 1 января 2000 года;
 - Для каждой книги посчитайте количество обзоров и среднюю оценку;
 - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
 - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
 - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Как выполнить задание?**
 - Опишите цели исследования;
 - Исследуйте таблицы — выведите первые строки;
 - Сделайте по одному SQL-запросу для решения каждого задания;
 - Выведите результаты каждого запроса в тетрадке;
 - Опишите выводы по каждой из решённых задач.

# Загрузка данных

In [1]:
# импорт библиотек
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string) 

In [2]:
books = pd.io.sql.read_sql('SELECT * FROM books', con = engine) 
authors = pd.io.sql.read_sql('SELECT * FROM authors', con = engine) 
publishers = pd.io.sql.read_sql('SELECT * FROM publishers', con = engine) 
ratings = pd.io.sql.read_sql('SELECT * FROM ratings', con = engine) 
reviews = pd.io.sql.read_sql('SELECT * FROM reviews', con = engine) 

In [3]:
pd.io.sql.read_sql('SELECT * FROM books LIMIT 5;', con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
pd.io.sql.read_sql('SELECT * FROM authors LIMIT 5;', con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [7]:
pd.io.sql.read_sql('SELECT * FROM publishers LIMIT 5;', con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [9]:
pd.io.sql.read_sql('SELECT * FROM ratings LIMIT 5;', con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [11]:
pd.io.sql.read_sql('SELECT * FROM reviews LIMIT 5;', con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Данные загружены, можно приступать к запросам

# Запрос №1: количество книг, вышедших после 1 января 2000 года

In [13]:

query = '''

SELECT COUNT(book_id)
FROM books
WHERE publication_date >= '2000-01-01'

'''

pd.io.sql.read_sql(query, con = engine)

count
0    821

Число книг из датасета, вышедших после 1 яневаря 2000 года, равно 821.

# Запрос №2: количество обзоров и средняя оценка для каждой книги

In [14]:
query = '''

SELECT b.book_id, 
       b.title, 
       COUNT(DISTINCT rv.review_id) AS cnt_reviews, 
       ROUND(AVG(r.rating),3) AS avg_rating
FROM books AS b

LEFT JOIN reviews AS rv
ON b.book_id = rv.book_id

LEFT JOIN  ratings AS r
ON b.book_id = r.book_id

GROUP BY b.book_id, b.title
ORDER BY cnt_reviews DESC

'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        963                                Water for Elephants            6   
2        734                                   The Glass Castle            6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4        695  The Curious Incident of the Dog in the Night-Time            6   
..       ...                                                ...          ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0         3.663  
1         3.977  
2         4.207  
3         4.415  
4         4.081  
..          ...  
995       3.667  
996       3.000  
997       5.000  
998       4.000  
999       4.000  

[1000 rows x 4 columns]

Для каждой книги посчитана средняя оценка и число обзоров.

# Запрос №3: определение издательства, выпустившего наибольшее количество книг толще 50 страниц

In [15]:
query3 = '''SELECT 
publishers.publisher AS publisher, 
COUNT(books.book_id) AS books_amount

FROM publishers 
INNER JOIN books ON books.publisher_id = publishers.publisher_id 

WHERE 
num_pages > 50

GROUP BY 
publishers.publisher 

ORDER BY 
books_amount DESC 

LIMIT 1;
'''
pd.io.sql.read_sql(query3, con = engine)

publisher  books_amount
0  Penguin Books            42

Издательство, которое выпустило наибольшее количество книг толще 50 страниц - это Penguin Books , и выпустило оно 42 книги

# Запрос №4: определение автора с самой высокой средней оценкой книг среди книг с 50 и большим числом оценок

In [16]:
query4 = '''
SELECT
SUBQ.author_id AS author_id,
authors.author AS author,
SUBQ.avg_rating AS avg_rating

FROM
    (
    SELECT
    books.book_id,
    books.author_id,
    AVG(ratings.rating) AS avg_rating
        
    FROM 
    books 
    INNER JOIN ratings ON ratings.book_id = books.book_id

    GROUP BY 
    books.book_id

    HAVING
    COUNT(ratings.rating_id) >= 50
    ) AS SUBQ
INNER JOIN authors ON authors.author_id = SUBQ.author_id

ORDER BY
SUBQ.avg_rating DESC

LIMIT 1;
'''
pd.io.sql.read_sql(query4, con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.414634

Было определено, что J.K. Rowling/Mary GrandPré являются авторами с самой высокой средней оценкой книг при рассмотрении книг датасета с числом оценок 50 и более.Средняя оценка книг авторов составила почти 4.4

# Запрос №5: расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [17]:
query5 = '''
SELECT
AVG(SUBQ1.reviews_amount) AS avg_reviews

FROM 
(
SELECT 
SUBQ.username AS username,
COUNT(reviews.review_id) AS reviews_amount 
        
FROM 
    (
    SELECT
    ratings.username AS username
    
    FROM
    ratings
    
    GROUP BY 
    ratings.username
    
    HAVING 
    COUNT(ratings.rating_id) > 50
    ) AS SUBQ 
INNER JOIN reviews ON reviews.username = SUBQ.username

GROUP BY
SUBQ.username
) AS SUBQ1;

'''

pd.io.sql.read_sql(query5, con = engine)

avg_reviews
0    24.333333

Выбрала пользователей, которые поставили более 50 оценок книгам, рассчитала, что среднее количетиво обзоров на книги от этих пользователей равно 24,3.

# Вывод

В данном исследовании были проанализированы датасеты, с информацией о книгах, их авторах, издателшям, а также с оценками и рейтингами пользователей.

Было выявлено, что из 1000 книг в датасете после 1 января 2000 года вышла 821 книга;

Для каждой книги была посчитана средняя оценка и число обзоров;

Было определено, что Penguin Books является издательством, выпустившим наибольшее количество книг (не считая брошюр), из 1000 книг в датасете 42 книги приходятся на издательство Penguin Books;

Было определено, что J.K. Rowling/Mary GrandPré являются авторами с самой высокой средней оценкой книг при рассмотрении книг датасета с числом оценок 50 и более. Средняя оценка книг авторов составила почти 4.4 из 5;

Также были выбраны пользователи, которые поставили более 50 оценок книгам, среди них было рассчитано среднее число обзоров, которые они делают, получилось 24,3 обзора на такого пользователя, их можно отнести к наиболее активным читателям в разрезе данного датасета.

На основании проведенного анализа можно делать специальные подборки книг для активных пользоватиелей, можно составить портрет читателя, посмотреть, какие авторы удостаиваются наибольших оценок, добавлять на платформу книги этих авторов.

Можно построить рейтинг издательств, чью книги также удостаиваются наилучших оценок и наибольшего количества обзоров, и пополнять коллекцию книг приложения книгами этих издательств.

Уже отмечены такие авторы как J.K. Rowling/Mary GrandPré и продуктивное издательство Penguin Books, на них нужно обратить внимание в первую очередь.